TheMET/TheMET/ViewModels/CultureArtsViewModel.swift

```diff
@@ -0,0 +1,319 @@
+            }
+        }
+    }
+    
+    func artWillBeDisplayed(artState: CultureArtState.LoadedArtState) {
+        if case .loaded = artState.imageStatus {
+            return
+        }
+        let artID = artState.objectID
+        guard !self.artsWithImageLoading.contains(artID) else { return }
+        if let art = self.loadedArts.first(where: {$0.objectID == artID}) {
+            guard let imageURL = art.primaryImage else {
+                return
+            }
+            self.artsWithImageLoading.append(art.objectID)
+            self.imageLoader.loadImage(urlString: imageURL) { [weak self] image in
+                self?.imageDidLoad(artID: art.objectID, image: image)
+                
+            }
+        }
+    }
+    
+    private func imageDidLoad(artID: Int, image: UIImage?) {
+        let imageStatus: LoadingStatus<UIImage>
+        if let image = image {
+            imageStatus = .loaded(image)
+        } else {
+            imageStatus = .failed(FavoritesLoadingError.imageNotLoaded)
+        }
+        switch self.cultureArtStatesList {
+        case .loaded(let statesList):
+            if let index = statesList.firstIndex(where: { cultureArtState in
+                switch cultureArtState {
+                case .placeholder(let placeholderArtID):
+                    return placeholderArtID == artID
+                case .loaded(let artState):
+                    return artState.objectID == artID
+                }
+            }) {
+                var loadedArt = statesList[index]
+                switch loadedArt{
+                case .placeholder:
+                    break
+                case .loaded(var loadedArtState):
+                    loadedArtState.imageStatus = imageStatus
+                    loadedArt = .loaded(artState: loadedArtState)
+                    var newArts: [CultureArtState] = statesList
+                    newArts[index] = loadedArt
+                    self.cultureArtStatesList = .loaded(newArts)
+                }
+            }
+        case .loading, .failed:
+            return
+        }
+        self.artsWithImageLoading.removeAll(where: { $0 == artID })
+    }
+    
+    private func removeLoadingArtId(artId: ArtID) {
+        self.loadingArtIds.removeAll { id in
+            return id == artId
+        }
+    }
+    
+    private func isArtLiked(artId: ArtID) -> Bool {
+        return self.favoriteService.favoriteArts.contains(where: { favoriteArt in
+            return favoriteArt.objectID == artId
+        })
+    }
+    
+    func likeButtonDidTap(artID: ArtID) {
+        if self.favoriteService.favoriteArts.contains(where: { $0.objectID == artID }) {
+            self.favoriteService.removeArt(id: artID)
+        } else {
+            if let art = self.loadedArts.first(where: {$0.objectID == artID}) {
+                self.favoriteService.addFavoriteArt(art)
+            }
+        }
+    }
+    
+    func tableViewWillDisplay(cultureArtState: CultureArtState) {
```

Comment:
Bad name. ViewModel must not know that the arts are displayed with TableView. Replace with `artWillBeDisplayed` or similar.